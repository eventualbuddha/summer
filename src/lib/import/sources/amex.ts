import { Result } from '@badrap/result';
import { DateTime } from 'luxon';
import type { ImportedTransaction } from '../ImportedTransaction';
import { StatementMetadata } from '../StatementMetadata';
import { parseAmount, ParseMoneyError } from '../parse/money';
import type { Statement } from '../statement/Statement';
import type { Page } from '../statement/page';

export type ImportStatementError = ParseMoneyError | ParseStatementSummaryError;

export class StatementSummary extends StatementMetadata {
	accountHolder: string;
	previousBalance: number;
	paymentsAndCredits: number;
	newCharges: number;
	fees: number;
	interestCharged: number;
	newBalance: number;
	minimumPaymentDue: number;
	creditLimit: number;
	availableCredit: number;
	cashAdvanceLimit: number;
	availableCash: number;

	constructor(
		closingDate: DateTime,
		account: string,
		accountHolder: string,
		previousBalance: number,
		paymentsAndCredits: number,
		newCharges: number,
		fees: number,
		interestCharged: number,
		newBalance: number,
		minimumPaymentDue: number,
		creditLimit: number,
		availableCredit: number,
		cashAdvanceLimit: number,
		availableCash: number
	) {
		super(closingDate, account);
		this.accountHolder = accountHolder;
		this.previousBalance = previousBalance;
		this.paymentsAndCredits = paymentsAndCredits;
		this.newCharges = newCharges;
		this.fees = fees;
		this.interestCharged = interestCharged;
		this.newBalance = newBalance;
		this.minimumPaymentDue = minimumPaymentDue;
		this.creditLimit = creditLimit;
		this.availableCredit = availableCredit;
		this.cashAdvanceLimit = cashAdvanceLimit;
		this.availableCash = availableCash;
	}
}

export class ParseStatementSummaryError extends Error {
	readonly pageNumber: number;
	readonly cause?: Error;

	constructor(pageNumber: number, message: string, cause?: Error) {
		super(`${message} on page ${pageNumber}${cause ? `: ${cause.message}` : ''}`);
		this.pageNumber = pageNumber;
		this.cause = cause;
	}
}

export async function* parseStatement(
	statement: Statement
): AsyncGenerator<Result<ImportedTransaction | StatementSummary, ImportStatementError>> {
	for (const page of statement.pages) {
		yield parseStatementSummary(page);
	}
}

export function parseStatementSummary(
	page: Page
): Result<StatementSummary, ParseStatementSummaryError> {
	const closingDateLabel = page.navigator.find('Closing Date');
	const closingDateValue = closingDateLabel?.findRight(/^\d+\/\d+\/\d+$/);

	if (!closingDateValue) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing closing date'));
	}

	const closingDate = DateTime.fromFormat(closingDateValue.text.str.trim(), 'MM/dd/yy');

	if (!closingDate.isValid) {
		return Result.err(
			new ParseStatementSummaryError(
				page.pageNumber,
				`Invalid closing date: ${closingDate.invalidExplanation}`
			)
		);
	}

	const accountHolderLabel = closingDateLabel?.findUp(/\w+/, { alignment: 'left', maxGap: 10 });

	if (!accountHolderLabel) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing account holder'));
	}

	const accountEndingLabel = closingDateLabel?.findDown(/Account Ending/, { alignment: 'left' });
	const accountEndingValue = accountEndingLabel?.findRight(/[-\d]+/);

	if (!accountEndingValue) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing account number'));
	}

	const previousBalanceLabel = page.navigator.find(/Previous Balance/);
	const entries = previousBalanceLabel?.findTable(
		[
			/Previous Balance/,
			/Payments\/Credits/,
			/New Charges/,
			/Fees/,
			/Interest Charged/,
			/New Balance/,
			/Minimum Payment Due/,
			/Credit Limit/,
			/Available Credit/,
			/Cash Advance Limit/,
			/Available Cash/
		],
		[/\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/],
		{ labelAlignment: 'left', valueAlignment: 'right' }
	);

	if (!entries) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing summary info'));
	}

	const account = accountEndingValue.text.str.trim();
	const accountHolder = accountHolderLabel.text.str.trim();
	const previousBalance = parseAmount(entries[0][1].text.str);
	const paymentsAndCredits = parseAmount(entries[1][1].text.str);
	const newCharges = parseAmount(entries[2][1].text.str);
	const fees = parseAmount(entries[3][1].text.str);
	const interestCharged = parseAmount(entries[4][1].text.str);
	const newBalance = parseAmount(entries[5][1].text.str);
	const minimumPaymentDue = parseAmount(entries[6][1].text.str);
	const creditLimit = parseAmount(entries[7][1].text.str);
	const availableCredit = parseAmount(entries[8][1].text.str);
	const cashAdvanceLimit = parseAmount(entries[9][1].text.str);
	const availableCash = parseAmount(entries[10][1].text.str);

	const result = Result.all([
		previousBalance,
		paymentsAndCredits,
		newCharges,
		fees,
		interestCharged,
		newBalance,
		minimumPaymentDue,
		creditLimit,
		availableCredit,
		cashAdvanceLimit,
		availableCash
	]);

	if (result.isErr) {
		return Result.err(
			new ParseStatementSummaryError(page.pageNumber, 'Failed to parse summary info', result.error)
		);
	}

	return Result.ok(new StatementSummary(closingDate, account, accountHolder, ...result.value));
}
