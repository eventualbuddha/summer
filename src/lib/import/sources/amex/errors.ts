import {
	InconsistentValueError,
	InvalidDateError,
	MissingHeaderError,
	MissingValueError
} from '../../parse/errors';
import { ParseMoneyError } from '../../parse/money';
import type { PageText } from '../../statement/page';
import { InvalidMoneyAmountError } from '../schwab';

export type ParseActivityEntriesErrorCause =
	| InvalidDateError
	| InvalidMoneyAmountError
	| ParseMoneyError
	| MissingHeaderError
	| MissingValueError
	| InconsistentValueError<number>;

export interface ParseActivityEntriesErrorContext {
	pageNumber?: number;
	contentText?: PageText;
	searchFromText?: PageText;
	searchDirection?: 'left' | 'right' | 'up' | 'down';
}

export class ParseActivityEntriesError extends Error {
	readonly cause: ParseActivityEntriesErrorCause;
	readonly context?: ParseActivityEntriesErrorContext;

	constructor(cause: ParseActivityEntriesErrorCause, context?: ParseActivityEntriesErrorContext) {
		super(cause.message);
		this.cause = cause;
		this.context = context;
	}
}
