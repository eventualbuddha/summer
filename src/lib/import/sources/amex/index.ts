export * from './charges';
export * from './credits';
export * from './statement';
export * from './summary';

export const id = 'american-express';
export const name = 'American Express';
