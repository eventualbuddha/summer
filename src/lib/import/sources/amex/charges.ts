import type { ImportedTransaction } from '$lib/import/ImportedTransaction';
import {
	InconsistentValueError,
	MissingHeaderError,
	MissingValueError
} from '$lib/import/parse/errors';
import { parseAmount } from '$lib/import/parse/money';
import type { StatementTextLocation } from '$lib/import/statement/navigation';
import { Result } from '@badrap/result';
import { DateTime } from 'luxon';
import { ParseActivityEntriesError } from './errors';

export class Charge implements ImportedTransaction {
	pageNumber: number;
	date: DateTime<true>;
	descriptionLines: string[];
	location: string;
	state: string;
	amount: number;

	constructor(
		pageNumber: number,
		date: DateTime<true>,
		descriptionLines: string[],
		location: string,
		state: string,
		amount: number
	) {
		this.pageNumber = pageNumber;
		this.date = date;
		this.descriptionLines = descriptionLines;
		this.location = location;
		this.state = state;
		this.amount = amount;
	}

	get statementDescription(): string {
		return this.descriptionLines.join('\n');
	}
}

export function parseNewCharges(
	sectionHeader: StatementTextLocation,
	nextSectionHeader: StatementTextLocation
): Result<Charge[], ParseActivityEntriesError> {
	const charges: Charge[] = [];

	const totalNewChargesLabel = sectionHeader.findAfter(/Total\s+New\s+Charges/);

	if (!totalNewChargesLabel || totalNewChargesLabel.isBelow(nextSectionHeader)) {
		return Result.err(
			new ParseActivityEntriesError(new MissingHeaderError('Total New Charges'), {
				pageNumber: (totalNewChargesLabel ?? sectionHeader).pageNumber,
				contentText: totalNewChargesLabel?.text,
				searchFromText: sectionHeader.text
			})
		);
	}

	const totalNewChargesValue = totalNewChargesLabel.findRight(/\d+/);

	if (!totalNewChargesValue || totalNewChargesValue.isBelow(nextSectionHeader)) {
		return Result.err(
			new ParseActivityEntriesError(new MissingValueError('Total New Charges'), {
				pageNumber: (totalNewChargesValue ?? totalNewChargesLabel).pageNumber,
				contentText: totalNewChargesValue?.text,
				searchFromText: totalNewChargesLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const totalNewChargesAmount = parseAmount(totalNewChargesValue.text.str);

	if (totalNewChargesAmount.isErr) {
		return Result.err(
			new ParseActivityEntriesError(totalNewChargesAmount.error, {
				pageNumber: (totalNewChargesValue ?? totalNewChargesLabel).pageNumber,
				contentText: totalNewChargesValue?.text,
				searchFromText: totalNewChargesLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const totalNewCharges = totalNewChargesAmount.value;

	let lastAmountHeader = totalNewChargesValue;
	while (lastAmountHeader.isAbove(nextSectionHeader)) {
		const nextAmountHeader = lastAmountHeader.findDown('Amount', {
			alignment: 'right'
		});

		if (!nextAmountHeader) {
			break;
		}

		lastAmountHeader = nextAmountHeader;

		let lastAmountValue = lastAmountHeader;
		while (lastAmountValue.isAbove(nextSectionHeader)) {
			const amountValue = lastAmountValue.findDown(/\d+/, { alignment: 'right' });

			if (amountValue?.isBelow(nextSectionHeader)) {
				lastAmountValue = amountValue;
				break;
			}

			if (!amountValue) {
				return Result.err(
					new ParseActivityEntriesError(new MissingValueError('Amount'), {
						pageNumber: (amountValue ?? lastAmountValue).pageNumber,
						searchFromText: lastAmountValue.text,
						searchDirection: 'down'
					})
				);
			}

			const parsedAmount = parseAmount(amountValue.text.str);

			if (parsedAmount.isErr) {
				return Result.err(
					new ParseActivityEntriesError(parsedAmount.error, {
						pageNumber: (amountValue ?? lastAmountValue).pageNumber,
						contentText: amountValue.text,
						searchFromText: lastAmountValue.text,
						searchDirection: 'down'
					})
				);
			}

			const stateValue = amountValue.findLeft(/\w+/);

			if (!stateValue || stateValue.isBelow(nextSectionHeader)) {
				return Result.err(
					new ParseActivityEntriesError(new MissingValueError('State'), {
						pageNumber: (stateValue ?? amountValue).pageNumber,
						contentText: stateValue?.text,
						searchFromText: amountValue.text,
						searchDirection: 'left'
					})
				);
			}

			const locationValue = stateValue.findLeft(/.{4,}/);

			if (!locationValue || locationValue.isBelow(nextSectionHeader)) {
				return Result.err(
					new ParseActivityEntriesError(new MissingValueError('Location'), {
						pageNumber: (locationValue ?? stateValue).pageNumber,
						contentText: locationValue?.text,
						searchFromText: stateValue.text,
						searchDirection: 'left'
					})
				);
			}

			const descriptionValue = locationValue.findLeft(/.{4,}/);

			if (!descriptionValue || descriptionValue.isBelow(nextSectionHeader)) {
				return Result.err(
					new ParseActivityEntriesError(new MissingValueError('Description'), {
						pageNumber: (descriptionValue ?? locationValue).pageNumber,
						contentText: descriptionValue?.text,
						searchFromText: locationValue.text,
						searchDirection: 'left'
					})
				);
			}

			const dateValue = descriptionValue.findLeft(/^\d\d\/\d\d\/\d\d/);

			if (!dateValue || dateValue.isBelow(nextSectionHeader)) {
				return Result.err(
					new ParseActivityEntriesError(new MissingValueError('Date'), {
						pageNumber: (dateValue ?? descriptionValue).pageNumber,
						contentText: dateValue?.text,
						searchFromText: descriptionValue.text,
						searchDirection: 'left'
					})
				);
			}

			const date = DateTime.fromFormat(dateValue.text.str, 'MM/dd/yy');

			if (!date.isValid) {
				return Result.err(
					new ParseActivityEntriesError(new MissingValueError('Date'), {
						pageNumber: (dateValue ?? descriptionValue).pageNumber,
						contentText: dateValue.text,
						searchFromText: descriptionValue.text,
						searchDirection: 'left'
					})
				);
			}

			charges.push(
				new Charge(
					dateValue.pageNumber,
					date,
					[descriptionValue.text.str],
					locationValue.text.str,
					stateValue.text.str,
					parsedAmount.value
				)
			);

			lastAmountValue = amountValue;
		}

		lastAmountHeader = nextAmountHeader;
	}

	const computedTotalChargesAmount = charges.reduce((sum, { amount }) => sum + amount, 0);

	if (computedTotalChargesAmount !== totalNewCharges) {
		return Result.err(
			new ParseActivityEntriesError(
				new InconsistentValueError('Total New Charges', totalNewCharges, computedTotalChargesAmount)
			)
		);
	}

	return Result.ok(charges);
}
