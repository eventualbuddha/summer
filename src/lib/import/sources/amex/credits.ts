import type { ImportedTransaction } from '$lib/import/ImportedTransaction';
import { ParseStatementError } from '$lib/import/parse/errors';
import { parseAmount } from '$lib/import/parse/money';
import type { StatementTextLocation } from '$lib/import/statement/navigation';
import { Result } from '@badrap/result';
import { DateTime } from 'luxon';

export class Credit implements ImportedTransaction {
	#pageNumber: number;
	date: DateTime<true>;
	card: string;
	descriptionLines: string[];
	amount: number;

	constructor(
		pageNumber: number,
		date: DateTime<true>,
		card: string,
		descriptionLines: string[],
		amount: number
	) {
		this.#pageNumber = pageNumber;
		this.date = date;
		this.card = card;
		this.descriptionLines = descriptionLines;
		this.amount = amount;
	}

	get pageNumber(): number {
		return this.#pageNumber;
	}

	get statementDescription(): string {
		return this.descriptionLines.join(' ');
	}
}

function parseCreditFromDate(
	dateValue: StatementTextLocation,
	endOfSection: StatementTextLocation
): Result<Credit, ParseStatementError> {
	const date = DateTime.fromFormat(dateValue.text.str.replace(/\*$/, ''), 'MM/dd/yy');

	if (!date.isValid) {
		return Result.err(
			ParseStatementError.InvalidValue(
				`Invalid date "${dateValue.text.str}": ${date.invalidExplanation}`,
				{
					pageNumber: dateValue.pageNumber,
					contentText: dateValue.text
				}
			)
		);
	}

	const cardValue = dateValue.findRight(/\w+/);

	if (!cardValue || cardValue.isBelow(endOfSection)) {
		return Result.err(
			ParseStatementError.MissingValue('Missing "card" value', {
				pageNumber: (cardValue ?? dateValue).pageNumber,
				contentText: cardValue?.text,
				searchFromText: dateValue.text,
				searchDirection: 'right'
			})
		);
	}

	const descriptionStart = cardValue.findRight(/\w+/);

	if (!descriptionStart || descriptionStart.isBelow(endOfSection)) {
		return Result.err(
			ParseStatementError.MissingValue('Missing "description" value', {
				pageNumber: (descriptionStart ?? cardValue).pageNumber,
				contentText: descriptionStart?.text,
				searchFromText: cardValue.text,
				searchDirection: 'right'
			})
		);
	}

	const descriptionLines = [descriptionStart];
	let lastDescriptionLine = descriptionStart;
	while (lastDescriptionLine.isAbove(endOfSection)) {
		const nextDescriptionLine = lastDescriptionLine.findDown(/\w+/, {
			alignment: 'left',
			maxGap: 10
		});
		if (!nextDescriptionLine) break;
		lastDescriptionLine = nextDescriptionLine;
		descriptionLines.push(lastDescriptionLine);
	}

	const amountValue = descriptionStart.findRight(/\d+\.\d{2}/);

	if (!amountValue || amountValue.isBelow(endOfSection)) {
		return Result.err(
			ParseStatementError.MissingValue('Missing "amount" value', {
				pageNumber: (amountValue ?? descriptionStart).pageNumber,
				contentText: amountValue?.text,
				searchFromText: descriptionStart.text,
				searchDirection: 'right'
			})
		);
	}

	const amount = parseAmount(amountValue.text.str);

	if (amount.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid amount "${amountValue.text.str}"`, {
				cause: amount.error,
				pageNumber: (amountValue ?? descriptionStart).pageNumber,
				contentText: amountValue.text,
				searchFromText: descriptionStart.text,
				searchDirection: 'right'
			})
		);
	}

	return Result.ok(
		new Credit(
			dateValue.pageNumber,
			date,
			cardValue.text.str,
			descriptionLines.map((line) => line.text.str),
			amount.value
		)
	);
}

function parseCreditsFromDateAlignedHeader(
	dateAlignedHeader: StatementTextLocation,
	endOfSection: StatementTextLocation
): Result<Credit[], ParseStatementError> {
	let lastDateAlignedLocation = dateAlignedHeader;
	const credits: Credit[] = [];

	while (true) {
		const dateValue = lastDateAlignedLocation.findDown(/^\d\d\/\d\d\/\d\d/, {
			alignment: 'left'
		});

		if (!dateValue || dateValue.isBelow(endOfSection)) {
			break;
		}

		const result = parseCreditFromDate(dateValue, endOfSection);

		if (result.isErr) {
			return Result.err(result.error);
		}

		credits.push(result.value);
		lastDateAlignedLocation = dateValue;
	}

	return Result.ok(credits);
}

export function parseCredits(
	sectionHeader: StatementTextLocation,
	nextSectionHeader: StatementTextLocation
): Result<Credit[], ParseStatementError> {
	const credits: Credit[] = [];

	const totalHeader = sectionHeader.findAfter('Total');

	if (!totalHeader || totalHeader.isBelow(nextSectionHeader)) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Total" label', {
				pageNumber: sectionHeader.pageNumber,
				contentText: totalHeader?.text,
				searchFromText: sectionHeader.text
			})
		);
	}

	const paymentsLabel = totalHeader.findAfter('Payments');

	if (!paymentsLabel || paymentsLabel.isBelow(nextSectionHeader)) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Payments" label', {
				pageNumber: (paymentsLabel ?? totalHeader).pageNumber,
				contentText: paymentsLabel?.text,
				searchFromText: totalHeader.text
			})
		);
	}

	const paymentsValue = totalHeader.findDown(/\d+/, {
		alignment: 'right',
		maxGap: 20
	});

	if (!paymentsValue || paymentsValue.isBelow(nextSectionHeader)) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Payments" label', {
				pageNumber: (paymentsValue ?? totalHeader).pageNumber,
				contentText: paymentsValue?.text,
				searchFromText: totalHeader.text,
				searchDirection: 'down'
			})
		);
	}

	const paymentsAmount = parseAmount(paymentsValue.text.str);

	if (paymentsAmount.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid amount: "${paymentsValue.text.str}"`, {
				cause: paymentsAmount.error,
				pageNumber: paymentsValue.pageNumber,
				contentText: paymentsValue.text,
				searchFromText: totalHeader.text,
				searchDirection: 'down'
			})
		);
	}

	const totalPaymentsAndCreditsLabel = paymentsLabel.findDown('Total Payments and Credits', {
		alignment: 'left'
	});

	if (!totalPaymentsAndCreditsLabel || totalPaymentsAndCreditsLabel.isBelow(nextSectionHeader)) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Total Payments and Credits" label', {
				pageNumber: (totalPaymentsAndCreditsLabel ?? paymentsLabel).pageNumber,
				contentText: totalPaymentsAndCreditsLabel?.text,
				searchFromText: paymentsLabel.text,
				searchDirection: 'down'
			})
		);
	}

	const totalPaymentsAndCreditsValue = totalHeader.findDown(
		(text) => text.isHorizontallyAlignedWith(totalPaymentsAndCreditsLabel.text),
		{ alignment: 'right' }
	);

	if (!totalPaymentsAndCreditsValue || totalPaymentsAndCreditsValue.isBelow(nextSectionHeader)) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Total Payments and Credits" label', {
				pageNumber: (totalPaymentsAndCreditsValue ?? totalHeader).pageNumber,
				contentText: totalPaymentsAndCreditsValue?.text,
				searchFromText: totalHeader.text,
				searchDirection: 'down'
			})
		);
	}

	const totalPaymentsAndCreditsAmount = parseAmount(totalPaymentsAndCreditsValue.text.str);

	if (totalPaymentsAndCreditsAmount.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(
				`Invalid amount: "${totalPaymentsAndCreditsValue.text.str}"`,
				{
					cause: totalPaymentsAndCreditsAmount.error,
					pageNumber: totalPaymentsAndCreditsValue.pageNumber,
					contentText: totalPaymentsAndCreditsValue.text,
					searchFromText: totalPaymentsAndCreditsLabel.text,
					searchDirection: 'down'
				}
			)
		);
	}

	const creditsAmount = totalPaymentsAndCreditsAmount.value - paymentsAmount.value;

	if (paymentsAmount.value !== 0) {
		const paymentsHeader = totalPaymentsAndCreditsLabel.findDown('Payments', {
			alignment: 'left'
		});

		if (!paymentsHeader || paymentsHeader.isBelow(nextSectionHeader)) {
			return Result.err(
				ParseStatementError.MissingLabel('Missing "Payments and Credits: Payments" label', {
					pageNumber: (paymentsHeader ?? totalPaymentsAndCreditsLabel).pageNumber,
					contentText: paymentsHeader?.text,
					searchFromText: totalPaymentsAndCreditsLabel.text,
					searchDirection: 'down'
				})
			);
		}

		const creditsHeader = paymentsHeader.findDown('Credits', {
			alignment: 'left'
		});

		const endOfPayments = creditsAmount === 0 ? nextSectionHeader : creditsHeader;

		if (!endOfPayments || endOfPayments.isBelow(nextSectionHeader)) {
			return Result.err(
				ParseStatementError.MissingLabel('Missing "Payments and Credits: Credits" label', {
					pageNumber: (endOfPayments ?? paymentsHeader).pageNumber,
					contentText: endOfPayments?.text,
					searchFromText: paymentsHeader.text,
					searchDirection: 'down'
				})
			);
		}

		const parsePaymentsResult = parseCreditsFromDateAlignedHeader(paymentsHeader, endOfPayments);

		if (parsePaymentsResult.isErr) {
			return Result.err(parsePaymentsResult.error);
		}

		credits.push(...parsePaymentsResult.value);
	}

	if (creditsAmount !== 0) {
		const creditsHeader = totalPaymentsAndCreditsLabel.findDown('Credits', {
			alignment: 'left'
		});

		if (!creditsHeader || creditsHeader.isBelow(nextSectionHeader)) {
			return Result.err(
				ParseStatementError.MissingLabel('Missing "Payments and Credits: Credits" label', {
					pageNumber: (creditsHeader ?? totalPaymentsAndCreditsLabel).pageNumber,
					contentText: creditsHeader?.text,
					searchFromText: totalPaymentsAndCreditsLabel.text,
					searchDirection: 'down'
				})
			);
		}

		const parsePaymentsResult = parseCreditsFromDateAlignedHeader(creditsHeader, nextSectionHeader);

		if (parsePaymentsResult.isErr) {
			return Result.err(parsePaymentsResult.error);
		}

		credits.push(...parsePaymentsResult.value);
	}

	return Result.ok(credits);
}
