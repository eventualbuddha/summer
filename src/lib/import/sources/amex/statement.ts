import { Result } from '@badrap/result';
import type { ImportedTransaction } from '../../ImportedTransaction';
import { ParseStatementError } from '../../parse/errors';
import type { Statement } from '../../statement/Statement';
import { parseNewCharges } from './charges';
import { parseCredits } from './credits';
import type { StatementSummary } from './summary';

export function* parseStatement(
	statement: Statement
): Generator<Result<ImportedTransaction | StatementSummary, ParseStatementError>> {
	const paymentsAndCreditsSectionHeader = statement.navigator.find('Payments and Credits');

	if (!paymentsAndCreditsSectionHeader) {
		yield Result.err(ParseStatementError.MissingLabel('Payments and Credits'));
		return;
	}

	const paymentsAndCreditsSectionSummaryLabel = paymentsAndCreditsSectionHeader.findDown(
		'Summary',
		{
			alignment: 'left',
			maxGap: 30
		}
	);

	if (!paymentsAndCreditsSectionSummaryLabel) {
		yield Result.err(
			ParseStatementError.MissingLabel('Payments and Credits: Summary', {
				pageNumber: paymentsAndCreditsSectionHeader.pageNumber,
				searchFromText: paymentsAndCreditsSectionHeader.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const paymentsAndCreditsDetailHeader = paymentsAndCreditsSectionSummaryLabel.findDown('Detail', {
		alignment: 'left'
	});

	if (!paymentsAndCreditsDetailHeader) {
		yield Result.err(
			ParseStatementError.MissingLabel('Payments and Credits: Detail', {
				pageNumber: paymentsAndCreditsSectionSummaryLabel.pageNumber,
				searchFromText: paymentsAndCreditsSectionSummaryLabel.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const newChargesSectionHeader = paymentsAndCreditsSectionHeader.findDown('New Charges', {
		alignment: 'left'
	});
	const newChargesSectionSummaryHeader = newChargesSectionHeader?.findDown('Summary', {
		alignment: 'left',
		maxGap: 30
	});

	if (!newChargesSectionHeader) {
		yield Result.err(
			ParseStatementError.MissingLabel('New Charges', {
				pageNumber: paymentsAndCreditsSectionHeader.pageNumber,
				searchFromText: paymentsAndCreditsSectionHeader.text,
				searchDirection: 'down'
			})
		);
		return;
	}
	if (!newChargesSectionSummaryHeader) {
		yield Result.err(
			ParseStatementError.MissingLabel('New Charges: Summary', {
				pageNumber: paymentsAndCreditsSectionHeader.pageNumber,
				searchFromText: paymentsAndCreditsSectionHeader.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const feesSectionHeader = newChargesSectionHeader.findDown('Fees', {
		alignment: 'left'
	});

	if (!feesSectionHeader) {
		yield Result.err(
			ParseStatementError.MissingLabel('New Charges: Fees', {
				pageNumber: newChargesSectionHeader.pageNumber,
				searchFromText: newChargesSectionHeader.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const interestChargedSectionHeader = feesSectionHeader.findDown('Interest Charged', {
		alignment: 'left'
	});

	if (!interestChargedSectionHeader) {
		yield Result.err(
			ParseStatementError.MissingLabel('New Charges: Interest Charged', {
				pageNumber: feesSectionHeader.pageNumber,
				searchFromText: feesSectionHeader.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const parseCreditsResult = parseCredits(paymentsAndCreditsSectionHeader, newChargesSectionHeader);

	if (parseCreditsResult.isErr) {
		yield Result.err(parseCreditsResult.error);
		return;
	}

	for (const credit of parseCreditsResult.value) {
		yield Result.ok(credit);
	}

	const parseChargesResult = parseNewCharges(newChargesSectionHeader, feesSectionHeader);

	if (parseChargesResult.isErr) {
		yield Result.err(parseChargesResult.error);
		return;
	}

	for (const charge of parseChargesResult.value) {
		yield Result.ok(charge);
	}
}
