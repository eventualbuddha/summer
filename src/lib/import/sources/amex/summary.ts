import { parseAmount } from '$lib/import/parse/money';
import type { Page } from '$lib/import/statement/page';
import { Result } from '@badrap/result';
import { DateTime } from 'luxon';
import { ParseStatementSummaryError } from '../../parse/errors';
import { StatementSummary } from './statement';

export function parseStatementSummary(
	page: Page
): Result<StatementSummary, ParseStatementSummaryError> {
	const closingDateLabel = page.navigator.find('Closing Date');
	const closingDateValue = closingDateLabel?.findRight(/^\d+\/\d+\/\d+$/);

	if (!closingDateValue) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing closing date'));
	}

	const closingDate = DateTime.fromFormat(closingDateValue.text.str.trim(), 'MM/dd/yy');

	if (!closingDate.isValid) {
		return Result.err(
			new ParseStatementSummaryError(
				page.pageNumber,
				`Invalid closing date: ${closingDate.invalidExplanation}`
			)
		);
	}

	const accountHolderLabel = closingDateLabel?.findUp(/\w+/, { alignment: 'left', maxGap: 10 });

	if (!accountHolderLabel) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing account holder'));
	}

	const accountEndingLabel = closingDateLabel?.findDown(/Account Ending/, { alignment: 'left' });
	const accountEndingValue = accountEndingLabel?.findRight(/[-\d]+/);

	if (!accountEndingValue) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing account number'));
	}

	const previousBalanceLabel = page.navigator.find(/Previous Balance/);
	const entries = previousBalanceLabel?.findTable(
		[
			/Previous Balance/,
			/Payments\/Credits/,
			/New Charges/,
			/Fees/,
			/Interest Charged/,
			/New Balance/,
			/Minimum Payment Due/,
			/Credit Limit/,
			/Available Credit/,
			/Cash Advance Limit/,
			/Available Cash/
		],
		[/\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/],
		{ labelAlignment: 'left', valueAlignment: 'right' }
	);

	if (!entries) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing summary info'));
	}

	const account = accountEndingValue.text.str.trim();
	const accountHolder = accountHolderLabel.text.str.trim();
	const previousBalance = parseAmount(entries[0][1].text.str);
	const paymentsAndCredits = parseAmount(entries[1][1].text.str);
	const newCharges = parseAmount(entries[2][1].text.str);
	const fees = parseAmount(entries[3][1].text.str);
	const interestCharged = parseAmount(entries[4][1].text.str);
	const newBalance = parseAmount(entries[5][1].text.str);
	const minimumPaymentDue = parseAmount(entries[6][1].text.str);
	const creditLimit = parseAmount(entries[7][1].text.str);
	const availableCredit = parseAmount(entries[8][1].text.str);
	const cashAdvanceLimit = parseAmount(entries[9][1].text.str);
	const availableCash = parseAmount(entries[10][1].text.str);

	const result = Result.all([
		previousBalance,
		paymentsAndCredits,
		newCharges,
		fees,
		interestCharged,
		newBalance,
		minimumPaymentDue,
		creditLimit,
		availableCredit,
		cashAdvanceLimit,
		availableCash
	]);

	if (result.isErr) {
		return Result.err(
			new ParseStatementSummaryError(page.pageNumber, 'Failed to parse summary info', result.error)
		);
	}

	return Result.ok(new StatementSummary(closingDate, account, accountHolder, ...result.value));
}
