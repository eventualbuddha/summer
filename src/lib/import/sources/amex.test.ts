import { expect, test } from 'vitest';
import * as fixtures from '../../../../tests/fixtures';
import { parseStatementSummary } from './amex';

test('checking account summary', async () => {
	expect(parseStatementSummary(fixtures.amex.skymiles.statement.pages[0]!)).toMatchSnapshot();
});
