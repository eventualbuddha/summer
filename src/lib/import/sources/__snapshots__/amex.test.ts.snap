// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checking account summary 1`] = `
t {
  "isErr": false,
  "isOk": true,
  "value": StatementSummary {
    "account": "9-39401",
    "accountHolder": "GINEVRA M WEASLEY",
    "accountName": undefined,
    "availableCash": 300000,
    "availableCredit": 1544432,
    "cashAdvanceLimit": 300000,
    "closingDate": "2025-02-21T00:00:00.000+00:00",
    "creditLimit": 2390000,
    "fees": 35000,
    "interestCharged": 0,
    "minimumPaymentDue": 6711,
    "newBalance": 111111,
    "newCharges": 382083,
    "paymentsAndCredits": -610845,
    "previousBalance": 610845,
  },
}
`;
